Many substrings of an unresolved rope shouldn't explode memory

kmiller68 opened the pull request:
#### 36b46ec7539e07337a21c998bd5435be4339f45c
<pre>
Many substrings of an unresolved rope shouldn&apos;t explode memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=291450">https://bugs.webkit.org/show_bug.cgi?id=291450</a>
<a href="https://rdar.apple.com/148899404">rdar://148899404</a>

Reviewed by Yusuke Suzuki.

In <a href="https://commits.webkit.org/284502@main">https://commits.webkit.org/284502@main</a> we added an optimization that would try to walk the rope string and look for the smallest rope to resolve.
Unfortunately, this means if you&apos;re taking many substrings of a much larger rope then we&apos;ll end up balloning memory usage. This happens for at
least one site. For their provided test case we went from ~7MB to ~55MB. In order to not regress that case we should resolve the outer most string
if we decide we need to do a resolution at all.

* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory-2.js: Added.
(globalThis.string_list.String.prototype.s):
(let.lo.lp):
(test):
* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory.js: Added.
(getLongRopeString):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::tryJSSubstringImpl):
(JSC::jsSubstring):

Canonical link: <a href="https://commits.webkit.org/293794@main">https://commits.webkit.org/293794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dacb1a147c20a060b46835334fae97df9d269c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105092 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50545 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19917 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28048 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76105 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33187 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102971 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56464 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8269 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49914 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92622 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84935 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8357 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107452 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98571 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19793 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27440 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86482 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/84587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21476 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29257 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6982 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20909 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32242 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122197 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26825 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34115 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30141 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSString.h:
```diff
@@ -985,70 +987,115 @@ ALWAYS_INLINE JSString* jsString(VM& vm, Ref<StringImpl>&& s)
 }
 
-inline JSString* jsSubstring(VM& vm, JSGlobalObject* globalObject, JSString* base, unsigned offset, unsigned length)
+inline JSString* JSString::substringImpl(VM& vm, JSGlobalObject* globalObject, unsigned offset, unsigned length)
```
How about naming it to trySubsubstringImpl to make it explicit that it can fail.

Constellation reviewed (approved):
r=me if JS2 is neutral

kmiller68 committed:
Many substrings of an unresolved rope shouldn't explode memory
https://bugs.webkit.org/show_bug.cgi?id=291450
rdar://148899404

Reviewed by Yusuke Suzuki.

In https://commits.webkit.org/284502@main we added an optimization that would try to walk the rope string and look for the smallest rope to resolve.
Unfortunately, this means if you're taking many substrings of a much larger rope then we'll end up balloning memory usage. This happens for at
least one site. For their provided test case we went from ~7MB to ~55MB. In order to not regress that case we should resolve the outer most string
if we decide we need to do a resolution at all.

* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory-2.js: Added.
(globalThis.string_list.String.prototype.s):
(let.lo.lp):
(test):
* JSTests/stress/many-substrings-of-rope-shouldnt-use-excessive-memory.js: Added.
(getLongRopeString):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::tryJSSubstringImpl):
(JSC::jsSubstring):

Canonical link: https://commits.webkit.org/293794@main
